File: gtci/17-modified-binary-search/cpp/6-minimum-difference-element.cpp
#include <cassert>
#include <cmath>
#include <iostream>
#include <vector>

class Solution {
public:
  static int searchMinDiffElement(const std::vector<int>& arr, int key) {
    if (arr.empty())
      return -1;
    
    int left = 0;
    int right = arr.size() - 1;

    while (left <= right) {
      const int mid = left + (right - left) / 2;

      if (arr[mid] == key) {
        return arr[mid];
      } else if (arr[mid] < key) {
        left = mid + 1;
      } else if (arr[mid] > key) {
        right = mid - 1;
      }
    }

    // after the loop:
    //  - left points to the largest element > key
    //  - right points to the smallest element < key
    if (left >= arr.size())
      return arr[right];
    if (right < 0)
      return arr[left];
    
    if (std::abs(arr[left] - key) < std::abs(arr[right] - key))
      return arr[left];
    else
      return arr[right];
  }
};

int main() {
    Solution sol;

    assert(sol.searchMinDiffElement({4, 6, 10}, 7) == 6);
    assert(sol.searchMinDiffElement({4, 6, 10}, 4) == 4);
    assert(sol.searchMinDiffElement({1, 3, 8, 10, 15}, 12) == 10);
    assert(sol.searchMinDiffElement({1, 3, 8, 10, 15}, 13) == 15);
    assert(sol.searchMinDiffElement({1, 3, 8, 10, 15}, -1) == 1);

    std::cout << "All test cases passed." << std::endl;
}

File: gtci/17-modified-binary-search/python3/6-minimum-difference-element.py
class Solution:
    def searchMinDiffElement(self, arr, key):
        if not arr:
            return -1

        left = 0
        right = len(arr) - 1

        min_difference_elem = None

        while left <= right:
            mid = (left + right) // 2

            if arr[mid] == key:
                return key
            elif arr[mid] < key:
                left = mid + 1
            elif arr[mid] > key:
                right = mid - 1

        # after the search:
        # - left is the largest element less than key
        # - right is the smallest element greater than key
        if left >= len(arr):
            return arr[right]

        if right < 0:
            return arr[left]

        if abs(arr[left] - key) < abs(arr[right] - key):
            return arr[left]
        else:
            return arr[right]


if __name__ == "__main__":
    sol = Solution()

    assert sol.searchMinDiffElement([4, 6, 10], 7) == 6
    assert sol.searchMinDiffElement([4, 6, 10], 4) == 4
    assert sol.searchMinDiffElement([1, 3, 8, 10, 15], 12) == 10
    assert sol.searchMinDiffElement([4, 6, 10], 17) == 10
    assert sol.searchMinDiffElement([4, 6, 10], -1) == 4
    assert sol.searchMinDiffElement([], 7) == -1

    print("All test cases passed.")
